### ISSUE TITLE
Allow vendored dependencies for `go_proto_library`

### ISSUE THREAD

**Issue by gkelly** [bazelbuild/rules_go#442]
As it stands, whenever `go_proto_repositories()` is invoked it automatically adds `new_go_repository` entries for `org_golang_x_net` and `org_golang_google_grpc`, regardless of if they exist or not yet.  The `go_proto_library` rule also adds dependencies on `@org_golang_x_net//context:go_default_library` and `@org_golang_google_grpc//:go_default_library` (if the proto has servies) and `@com_github_golang_protobuf//proto:go_default_library` even if it doesn't.

If the project using these rules has vendored `golang.org/x/net` or `google.golang.org/grpc` this can create version clashing which will cause compilation to fail due to conflicting object files or conflicting definitions during build.

I think allowing for overriding the dependency paths just like the rules allows overriding protoc and protoc_gen_go would allow for more flexibility for users.

**Comment by jayconrod** [comment]
I think the solution you propose is the right one: it should be possible to override the dependencies included automatically by `go_proto_library`.

It's possible to specify your own version of `x/net` and `grpc` by declaring them explicitly in WORKSPACE instead of using `go_proto_repositories`. However this still requires they be external (not vendored) and have specific names. This is unnecessarily inflexible.

**Comment by pmbethe09** [comment]
go_proto_repositories() is just a convenience method, which in your case should not be called.

However, as you note, deps are added including "@org_golang_x_net//context:go_default_library", which then won't hit your vendored version.

bazel *was* having rules depend on something in //external, so here say "//external:org_golang_x_net_context", and then somewhere someone adds an appropriate 'bind' statement.
However, then bazel folks were thinking to get rid of the often mis-used //external namespace, which is why we used the direct "@repo//" syntax in go_proto_library

So as @jayconrod suggests, maybe context and grpc deps should default to the above names, but go_proto_library should allow override?

**Comment by ianthehat** [comment]
The long term solution to this is nested workspaces, where we declare all the rules_go dependancies in the rules_go WORKSPACE, and then you can override them in your workspace just by having your own rules of the same name, but I think that is still a while from arriving.

**Comment by tomwilkie** [comment]
Is there a work around for this?  I'm currently getting:

```
inconsistent definition for type grpc.options during import
	struct { grpc.creds credentials.TransportCredentials; grpc.codec grpc.Codec; grpc.cp grpc.Compressor; grpc.dc grpc.Decompressor; grpc.maxMsgSize int; grpc.unaryInt grpc.UnaryServerInterceptor; grpc.streamInt grpc.StreamServerInterceptor; grpc.maxConcurrentStreams uint32; grpc.useHandlerImpl bool }
```

When doing builds.

**Comment by gkelly** [comment]
I guess we both had the same idea. :grin:

**Comment by jayconrod** [comment]
I think this was fixed in #442 (thanks @gkelly!), so I'll close this issue, but please reopen if anything else needs to be done.
